src/lib.rs

```diff
@@ -0,0 +1,54 @@
+#[macro_use]
+extern crate nickel;
+extern crate openssl;
+extern crate rand;
+extern crate rustc_serialize;
+extern crate byteorder;
+extern crate time;
+extern crate cookie;
+extern crate plugin;
+extern crate typemap;
+
+use std::fmt::Debug;
+use rustc_serialize::{Encodable, Decodable};
+use std::any::Any;
+use nickel::{Response, Middleware, MiddlewareResult, cookies, NickelError, Session, SessionStore};
+use nickel::status::StatusCode::Forbidden;
+
+pub type ValidateUserFunc<T> = Box<Fn(&T) -> bool + Send + Sync>;
+
+pub struct Authorizer<T, D> {
+    authorize: ValidateUserFunc<T>,
+    access_granted: Box<Middleware<D>+ Sized + Send + Sync>,
+}
+
+impl<T, D> Authorizer<T, D> {
+    pub fn new(authorize: ValidateUserFunc<T>,
+               access_granted: Box<Middleware<D> + Sized + Send + Sync>)
+               -> Authorizer<T, D> {
+        Authorizer { authorize: authorize, access_granted: access_granted }
+    }
+}
+
+impl< T, D: 'static> Middleware<D> for Authorizer<T, D>
+    where
+    D: AsRef<cookies::SecretKey> + SessionStore<Store=T>,
+    T: 'static + Any + Encodable + Decodable + Default + Debug
```

Comment:
Instead of having `ValidateUserFunc`, would an `Authorize` trait be useful, which we can bound `T` with?

e.g.
```rust
trait Authorize {
    type Permissions = bool;
    
    fn permission(&self) -> Self::Permissions;
}
```

Then the user could have more complicated permissions, e.g. `if let Some(Write) = env.current_user().permissions() { .. }`. I guess in my head there's still also a missing abstraction of `current_user` which is built upon `session` (as the session can store a lot of unrelated stuff, not user details).

Thinking further, then the `Authorizer` could be expressed something like `server.get(Authorize::only(Permissions::Owner, middleware! { "Secrets" }))`?